safenode/src/network/command.rs

```diff
@@ -0,0 +1,123 @@
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{
+    error::Error,
+    safe_msg::{SafeRequest, SafeResponse},
+    EventLoop,
+};
+use crate::network::error::Result;
+use futures::channel::oneshot;
+use libp2p::{multiaddr::Protocol, request_response::ResponseChannel, Multiaddr, PeerId};
+use std::collections::{hash_map, HashSet};
+use tracing::warn;
+use xor_name::XorName;
+
+/// Commands to send to the Swarm
+#[derive(Debug)]
+pub(crate) enum CmdToSwarm {
+    StartListening {
+        addr: Multiaddr,
+        sender: oneshot::Sender<Result<()>>,
+    },
+    Dial {
+        peer_id: PeerId,
+        peer_addr: Multiaddr,
+        sender: oneshot::Sender<Result<()>>,
+    },
+    StoreData {
+        xor_name: XorName,
+        sender: oneshot::Sender<Result<()>>,
+    },
+    GetDataProviders {
+        xor_name: XorName,
+        sender: oneshot::Sender<HashSet<PeerId>>,
+    },
+    SendSafeRequest {
+        req: SafeRequest,
+        peer: PeerId,
+        sender: oneshot::Sender<Result<SafeResponse>>,
+    },
+    SendSafeResponse {
+        resp: SafeResponse,
+        channel: ResponseChannel<SafeResponse>,
+    },
+}
+
+impl EventLoop {
+    pub(crate) fn handle_command(&mut self, command: CmdToSwarm) -> Result<(), Error> {
+        match command {
+            CmdToSwarm::StartListening { addr, sender } => {
+                let _ = match self.swarm.listen_on(addr) {
+                    Ok(_) => sender.send(Ok(())),
+                    Err(e) => sender.send(Err(e.into())),
+                };
+            }
+            CmdToSwarm::Dial {
+                peer_id,
+                peer_addr,
+                sender,
+            } => {
+                if let hash_map::Entry::Vacant(e) = self.pending_dial.entry(peer_id) {
+                    let _routing_update = self
+                        .swarm
+                        .behaviour_mut()
+                        .kademlia
+                        .add_address(&peer_id, peer_addr.clone());
+                    match self
+                        .swarm
+                        .dial(peer_addr.with(Protocol::P2p(peer_id.into())))
+                    {
+                        Ok(()) => {
+                            let _ = e.insert(sender);
+                        }
+                        Err(e) => {
+                            let _ = sender.send(Err(e.into()));
+                        }
+                    }
+                } else {
+                    warn!("Already dialing peer.");
+                }
+            }
+            CmdToSwarm::StoreData { xor_name, sender } => {
```

Comment:
Yeah, this doesn't contain the data, so `StoreData` confusing. `AnnounceHolding`, `RegisterAsHolder`, `RegisterAsProvider`, `StartProviding`etc. seems more accurate.

Reply:
Will be worked on in a follow up PR as stated earlier.